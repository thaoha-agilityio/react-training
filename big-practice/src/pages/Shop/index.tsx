import { shallow } from 'zustand/shallow';
import { useCallback, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { Button, Container, Flex, Spinner, Stack, Text, useDisclosure } from '@chakra-ui/react';

// Components
import { ConfirmModal, Products } from '@components';
import Banner from '@components/Banner';

// Constants
import { LIMIT_PRODUCTS, SHOP_CRUMBS, ROUTES, SUCCESS_MESSAGES, NO_RESULT } from '@constants';

// Hooks
import { useInfiniteProducts, useMutationDeleteProduct, useCustomToast } from '@hooks';

// Stores
import { useCartStore } from '@stores';

// Constants
import { IProduct, STATUSES } from '@types';

const Shop = (): JSX.Element => {
  const { isOpen, onOpen, onClose } = useDisclosure();
  const [selectedId, setSelectedId] = useState('');

  const {
    data: products,
    hasNextPage,
    isLoading: isLoadingProduct,
    isFetchingNextPage,
    fetchNextPage,
  } = useInfiniteProducts(LIMIT_PRODUCTS);

  const [addToCart] = useCartStore((state) => [state.addToCart], shallow);

  const { showToast } = useCustomToast();

  // Initialize navigate function
  const navigate = useNavigate();

  // Get the mutate from useMutationDeleteProduct hook
  const { mutate: deleteProduct, isLoading: isLoadingSubmit } = useMutationDeleteProduct();

  // Handle open Modal
  const handleOpen = useCallback((id: string) => {
    onOpen();
    setSelectedId(id);
  }, []);

  // handle Delete Item
  const handleDeleteItem = useCallback(() => {
    return deleteProduct(selectedId, {
      onError: (error) => {
        showToast(STATUSES.ERROR, error.message);
      },
      onSuccess: () => {
        onClose();
        showToast(STATUSES.SUCCESS, SUCCESS_MESSAGES.DELETED);
      },
    });
  }, [selectedId]);

  // Handle navigate to detail page
  const handleShowDetail = useCallback((id: string) => {
    navigate(ROUTES.DETAIL_PRODUCT_PARAMS + id);
  }, []);

  // Handle navigate to edit page
  const handleShowEditForm = useCallback((id: string) => {
    navigate(ROUTES.EDIT_PRODUCT_PARAMS + id);
  }, []);

  // Handle add product to cart
  const handleAddToCart = useCallback((product: IProduct) => {
    addToCart(product.id);
    showToast(STATUSES.SUCCESS, SUCCESS_MESSAGES.ADD_TO_CART);
  }, []);

  return (
    <>
      <Banner title='Shop' crumbs={SHOP_CRUMBS} />
      <Container maxW='container.xl'>
        <Stack m='auto' spacing='40px' py='30px'>
          {!products.length ? (
            <Text>{NO_RESULT.PRODUCTS}</Text>
          ) : (
            <>
              {/* TODO: update latter */}
              <Products
                products={products}
                isAction
                onAddToCart={handleAddToCart}
                onShowDetailItem={handleShowDetail}
                onEditItem={handleShowEditForm}
                onOpen={handleOpen}
              />

              {isLoadingProduct && <Spinner />}
              {hasNextPage && (
                <Flex justifyContent='center'>
                  <Button
                    w='245px'
                    h='48px'
                    isLoading={isFetchingNextPage}
                    onClick={() => fetchNextPage()}
                    disabled={!hasNextPage || isFetchingNextPage}
                  >
                    Show More
                  </Button>
                </Flex>
              )}
            </>
          )}
        </Stack>
      </Container>

      {isOpen && (
        <ConfirmModal
          isOpen={isOpen}
          title='Delete Confirmation'
          textCancel='Cancel'
          textSubmit='Yes, Delete'
          text='Are you sure you want to delete this item?'
          onClose={onClose}
          onSubmit={handleDeleteItem}
          isLoading={isLoadingSubmit}
        />
      )}
    </>
  );
};

export default Shop;
