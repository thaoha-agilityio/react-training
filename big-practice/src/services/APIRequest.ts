import axios, { AxiosError, AxiosResponse } from 'axios';

export const getData = async <T>(url: string): Promise<T> => {
  try {
    const res: AxiosResponse<T> = await axios.get<T>(url);

    return res.data;
  } catch (error) {
    throw error as AxiosError;
  }
};

export const postData = async <T>({ item, url }: { item: T; url: string }): Promise<T> => {
  try {
    const response = await axios.post<T>(url, item);

    return response.data;
  } catch (error) {
    throw error as AxiosError;
  }
};

export const putData = async <T>({ item, url }: { item: Partial<T>; url: string }): Promise<T> => {
  try {
    const response = await axios.put<T>(url, item);

    return response.data;
  } catch (error) {
    throw error as AxiosError;
  }
};

export const deleteData = async <T>(url: string): Promise<T> => {
  try {
    const response = await axios.delete<T>(url);

    return response.data;
  } catch (error) {
    throw error as AxiosError;
  }
};

const api = { getData, postData, putData };

export { api };
