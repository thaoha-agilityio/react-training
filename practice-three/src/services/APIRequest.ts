import axios from "axios";

const getData = async <T>(url: string): Promise<T | any> => {
  try {
    const res = await axios.get<T>(url);

    return {
      data: res.data,
      error: "",
    };
  } catch (error: any) {
    return {
      data: [],
      error: error,
    };
  }
};

const api = { getData };

export { api };
