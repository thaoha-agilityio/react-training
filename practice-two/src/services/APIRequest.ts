import axios from './config';

const getData = <T>(url: string, params?: object) => axios.get<never, T>(url, params);

export { getData };
