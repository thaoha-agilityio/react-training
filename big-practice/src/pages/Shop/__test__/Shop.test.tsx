import { act, fireEvent, renderHook } from '@testing-library/react';
import { useDisclosure } from '@chakra-ui/react';

//  Constants
import { MOCK_CARTS, MOCK_PRODUCTS, NO_RESULT, ROUTES } from '@constants';

// Hooks
import { renderWithRouterAndQuery } from '@helpers';

//  Services
import { api } from '@services/APIRequest';

// Stores
import * as stores from '@stores';

// Hooks
import * as hooks from '@hooks';

import Shop from '..';

// Mock the useNavigate hook
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

// Mock services
jest.mock('@services/APIRequest', () => ({
  __esModule: true,
  ...jest.requireActual('@services/APIRequest'),
}));

// Mock Hooks
jest.mock('@hooks', () => ({
  __esModule: true,
  ...jest.requireActual('@hooks'),
}));

describe('Shop component', () => {
  beforeEach(() => {
    jest.spyOn(api, 'getData').mockResolvedValue([MOCK_PRODUCTS]);

    (jest.spyOn(hooks, 'useInfiniteProducts') as jest.Mock).mockImplementation(() => ({
      data: [MOCK_PRODUCTS],
      isFetching: false,
      hasNextPage: true,
      fetchNextPage: jest.fn(),
    }));

    stores.useCartStore.setState({
      cart: MOCK_CARTS,
      addToCart: jest.fn(),
    });
  });

  it('should render component correctly', () => {
    const { container } = renderWithRouterAndQuery(<Shop />);

    expect(container).toBeInTheDocument();
  });

  it('should render message if the list is empty', () => {
    (jest.spyOn(hooks, 'useInfiniteProducts') as jest.Mock).mockImplementation(() => ({
      data: [],
    }));
    const { getByText } = renderWithRouterAndQuery(<Shop />);

    expect(getByText(NO_RESULT.PRODUCTS)).toBeInTheDocument();
  });

  it('should navigates to detail page when a detail button is clicked', () => {
    const { getAllByTestId } = renderWithRouterAndQuery(<Shop />);

    const detailBtn = getAllByTestId('detail-btn');
    fireEvent.click(detailBtn[0]);
    expect(mockNavigate).toHaveBeenCalledWith(ROUTES.DETAIL_PRODUCT_PARAMS + MOCK_PRODUCTS[0].id);
  });

  it('should navigates to edit page when a edit button is clicked', () => {
    const { getAllByTestId } = renderWithRouterAndQuery(<Shop />);

    const editBtn = getAllByTestId('edit-btn');
    fireEvent.click(editBtn[0]);

    expect(mockNavigate).toHaveBeenCalledWith(ROUTES.EDIT_PRODUCT_PARAMS + MOCK_PRODUCTS[0].id);
  });

  it('should call handleDeleteItem success when delete button is clicked', async () => {
    jest.spyOn(api, 'deleteData').mockResolvedValue('1');
    const { getAllByTestId, getByText } = renderWithRouterAndQuery(<Shop />);

    act(() => {
      // Click deleteBtn
      const deleteBtn = getAllByTestId('delete-btn');
      fireEvent.click(deleteBtn[0]);
    });

    act(() => {
      const submitBtn = getByText('Yes, Delete');
      fireEvent.click(submitBtn);
    });
  });

  it('should call handleDeleteItem error when delete button is clicked', async () => {
    jest.spyOn(api, 'deleteData').mockRejectedValue(new Error('Error'));

    const { getAllByTestId, getByText } = renderWithRouterAndQuery(<Shop />);

    act(() => {
      // Click deleteBtn
      const deleteBtn = getAllByTestId('delete-btn');
      fireEvent.click(deleteBtn[0]);
    });

    const submitBtn = getByText('Yes, Delete');
    act(() => {
      fireEvent.click(submitBtn);
    });
  });

  it('should call handleAddToCart when add to cart button is clicked', async () => {
    const { getAllByText } = renderWithRouterAndQuery(<Shop />);
    act(() => {
      // Click deleteBtn
      const addToCartBtn = getAllByText('Add to cart');
      fireEvent.click(addToCartBtn[0]);
    });
  });

  it('should update cart when existing product', async () => {
    const { getAllByText, getByText } = renderWithRouterAndQuery(<Shop />);
    act(() => {
      // Click deleteBtn
      const addToCartBtn = getAllByText('Add to cart');
      fireEvent.click(addToCartBtn[3]);
    });

    const showMoreBtn = getByText('Show More');
    fireEvent.click(showMoreBtn);
  });

  it('should render ConfirmModal component', () => {
    const { result } = renderHook(() => useDisclosure());
    const { getAllByTestId, getByText } = renderWithRouterAndQuery(<Shop />);

    const deleteBtn = getAllByTestId('delete-btn');
    fireEvent.click(deleteBtn[0]);
    expect(result.current.onOpen).toBeTruthy();

    const cancelBtn = getByText('Cancel');
    fireEvent.click(cancelBtn);
  });
});
