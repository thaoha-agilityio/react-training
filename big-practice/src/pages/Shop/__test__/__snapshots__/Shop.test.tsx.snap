// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shop component should render component correctly 1`] = `<div />`;
